include/El/core/Profiling.hpp

```diff
@@ -0,0 +1,248 @@
+    case Color::PURPLE_HEART: return "PURPLE_HEART";
+    case Color::MANGO_TANGO: return "MANGO_TANGO";
+    case Color::DARK_RED: return "DARK_RED";
+    case Color::EUCALYPTUS: return "EUCALYPTUS";
+    case Color::SAN_MARINO: return "SAN_MARINO";
+    }
+    return "UNKNOWN";
+}
+
+/** \brief Circularly loops through a list of colors returning a new
+ *      color each call.
+ *
+ *  While not guaranteed to remain this way, the is currently 20
+ *  colors, 19 unique with one repeat. The list is here:
+ *  http://there4.io/2012/05/02/google-chart-color-list/
+ *
+ *  The current color is currently an atomic size_t with relaxed
+ *  memory ordering, so multiple threads could see the same value. I
+ *  don't see this as an issue since the colors are just sugar on a
+ *  visualization tool.
+ *
+ *  \return A color.
+*/
+
+Color GetNextProfilingColor() noexcept;
+
+/** \brief Begin a profiling region.
+ *
+ *  \param desc A name or description for the profiling region.
+ *  \param color The color for the region.
+ */
+void BeginRegionProfile(char const *desc, Color color) noexcept;
+
+/** \brief End a profiling region
+ *
+ *  \param desc The name or description for the profiling region to be
+ *      stopped. Must match the one given when the region was started.
+ */
+void EndRegionProfile(const char *desc) noexcept;
+
+/** \brief Synchronize and begin a profiling region
+ *
+ *  \param desc A name or description for the profiling region.
+ *  \param color The color for the region.
+ *  \param si The SyncInfo object to be synchronized.
+ */
+template <Device D>
+void BeginRegionProfile(
+    char const *desc, Color color, SyncInfo<D> const& si) noexcept
+{
+    Synchronize(si);
+    BeginRegionProfile(desc, color);
+}
+
+/** \brief Syncrhonize and end a profiling region
+ *
+ *  \param desc The name or description for the profiling region to be
+ *      stopped. Must match the one given when the region was started.
+ *  \param si The SyncInfo object to be synchronized.
+ */
+template <Device D>
+void EndRegionProfile(const char* desc, SyncInfo<D> const& si) noexcept
+{
+    Synchronize(si);
+    EndRegionProfile(desc);
+}
+
+// RAII profiler.
+struct ProfileRegion
+{
+    ProfileRegion(std::string desc, Color color) noexcept
+        : desc_{std::move(desc)}
+    {
+        BeginRegionProfile(desc_.c_str(), color);
+    }
+    ~ProfileRegion() noexcept
+    {
+        EndRegionProfile(desc_.c_str());
+    }
+
+    // Disable copy
+    ProfileRegion(ProfileRegion const&) = delete;
+    ProfileRegion& operator=(ProfileRegion const&) = delete;
+
+    // Allow move -- allows the Make function to work
+    ProfileRegion(ProfileRegion&&) noexcept = default;
+    ProfileRegion& operator=(ProfileRegion&&) noexcept = default;
+
+    std::string desc_;
+};// struct ProfileRegion
+
+inline ProfileRegion MakeProfileRegion(std::string desc, Color color)
+{
+    return ProfileRegion(std::move(desc), color);
+}
+
+// Synchronous RAII profiler.
```

Comment:
It might be good to document what makes this "synchronous", since it's not obvious.